src/common/libhostlist/hostlist.c

```diff
@@ -0,0 +1,1140 @@
+        return;
+    }
+    /*  Case 2: current points to the hostrange which was deleted.
+     *  Simply reset depth to -1 so that hostlist_next() starts at the
+     *   next hostrange in the array.
+     */
+    if (index == c->index) {
+        c->depth = -1;
+        return;
+    }
+    /*  Case 3: hostrange elements were deleted at a position < c->index:
+     *  In this case, adjust current index back by the number of deleted
+     *   elements. (ensure index is at least -1 however.)
+     */
+    if (index < c->index) {
+        /*  Move index up by the number of deleted ranges */
+        if ((c->index -= n) < 0)
+            c->index = -1;
+    }
+    /* O/w current iterator should not be affected.
+     */
+}
+
+
+
+/* Delete the range at position n in the range array
+ * Assumes the hostlist lock is already held.
+ */
+static void hostlist_delete_range (struct hostlist *hl, int n)
+{
+    int i;
+    struct hostrange * old;
+
+    assert(hl != NULL);
+    assert(n < hl->nranges && n >= 0);
+
+    old = hl->hr[n];
+    for (i = n; i < hl->nranges - 1; i++)
+        hl->hr[i] = hl->hr[i + 1];
+    hl->nranges--;
+    hl->hr[hl->nranges] = NULL;
+
+    hostlist_shift_current (hl, n, 0, 1);
+
+    /* XXX caller responsible for adjusting nhosts */
+    /*hl->nhosts -= hostrange_count(old) */
+
+    hostrange_destroy (old);
+}
+
+
+/* Grab a single range from str
+ * returns 0 if str contained a valid number or range,
+ *        -1 if conversion of str to a range failed.
+ */
+static int parse_next_range (const char *str, struct _range *range)
+{
+    char *p, *q;
+    char *orig = strdup (str);
+    if (!orig)
+        return -1;
+
+    if ((p = strchr (str, '-'))) {
+        *p++ = '\0';
+        if (*p == '-') {  /* don't allow negative numbers */
+            errno = EINVAL;
+            goto error;
+        }
+    }
+    range->lo = strtoul (str, &q, 10);
+    if (q == str)  {
+        errno = EINVAL;
+        goto error;
+    }
+
+    range->hi = (p && *p) ? strtoul (p, &q, 10) : range->lo;
+
+    errno = EINVAL;
+    if (q == p || *q != '\0')
+        goto error;
+
+    if (range->lo > range->hi)
+        goto error;
+
+    if (range->hi - range->lo + 1 > MAX_RANGE ) {
+        errno = ERANGE;
+        goto error;
+    }
+
+    free (orig);
+    range->width = strlen (str);
+    return 0;
+
+  error:
+    free(orig);
+    return -1;
+}
+
+static int hostlist_append_host (struct hostlist *hl, const char *str)
+{
+    struct hostrange * hr;
+    struct hostname * hn;
+
+    assert (hl != NULL);
+
+    if (str == NULL || *str == '\0')
+        return 0;
+
+    if (!(hn = hostname_create (str)))
+        return -1;
+
+    if (hostname_suffix_is_valid (hn)) {
+        hr = hostrange_create (hn->prefix,
+                               hn->num,
+                               hn->num,
+                               hostname_suffix_width(hn));
+    } else
+        hr = hostrange_create_single (str);
+
+    hostlist_append_range (hl, hr);
```

Comment:
check for `hostrange_create()` / `hostrange_create_single()` error
check for `hostlist_append_range()` error